### PR TITLE
VQVAE and PixelCNN on OASIS dataset.

### PR DESCRIPTION
**Pull request by MattPChoy**
Hi, I've completed my project, which is implementing VQVAE and PixelCNN on the OASIS dataset. I've additionally added documentation about the project, including the intent of each file, the underlying theory and the results obtained in the README.md file. Please let me know if there's any information that I'm missing or if there's anything I can fix up. 

Thanks,
Matt Choy.

# Files of Interest
What files to look at, and what purpose they serve.
- [modules.py](https://github.com/MattPChoy/PatternFlow/blob/topic-recognition/recognition/MattChoy-VQVAE/modules.py) Contains module definitions for the VQVAE (& Encoder, Decoder) as well as PixelCNN
- [train.py](https://github.com/MattPChoy/PatternFlow/blob/topic-recognition/recognition/MattChoy-VQVAE/train.py) Uses the modules defined in `modules.py` to train and save the VQVAE and PixelCNN models.
- [dataset.py](https://github.com/MattPChoy/PatternFlow/blob/topic-recognition/recognition/MattChoy-VQVAE/dataset.py) Loads and normalises dataset.
- [constants.py](https://github.com/MattPChoy/PatternFlow/blob/topic-recognition/recognition/MattChoy-VQVAE/constants.py) Defines constants and hyperparameters to ensure that this model can be adapted to other datasets more easily.
- [predict.py](https://github.com/MattPChoy/PatternFlow/blob/topic-recognition/recognition/MattChoy-VQVAE/predict.py) Loads the trained VQVAE and PixelCNN models located in `./vqvae` and `./pixel_cnn` respectively to generate OASIS-like brain scan images.
- [util.py](https://github.com/MattPChoy/PatternFlow/blob/topic-recognition/recognition/MattChoy-VQVAE/util.py) Defines helper methods for visualisation, data processing etc.

## Table of Contents
- [Algorithm Description](#description-of-algorithm)
- [How Does it Work?](#how-does-it-work)
- [Inputs, Outputs and Algorithm Performance](#inputs-outputs-and-algorithm-performance)
- [Implementation Details](#implementation-details)
  - [Dependencies](#dependencies)
- [Dataset](#dataset)
- [Data Pre-Processing](#data-pre-processing)
- [Training, Validation and Testing Splits](#training-validation-and-testing-splits)

### CODE REVIEW

**Comment by MattPChoy** [comment]
Hi, I've noticed some formatting errors on my README.md file, but have amended them in [this commit](https://github.com/MattPChoy/PatternFlow/commit/261996eb3425d33e4718b8c925459741a6293718) but I don't seem to be able to merge them. Please let me know if you are able to do this, or if I'm able to do this on my end. Thanks.

**Comment by SiyuLiu0329** [comment]
**This is an initial inspection, no action is required at this point**
- commit messages: OK
- reconstruction: OK
- image gen: OK
- code commenting: OK
- readme: seems: OK

Extra node: very good

**Comment by shakes76** [comment]
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate use and implementation
Good spacing and comments
Header blocks
## Recognition Problem 
Solves problem
Driver Script present
File structure present
Shows Usage & Demo & Visualisation & Data usage 
Module present
Commenting
No Data leakage
Difficulty: Hard
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
ReadMe acceptable/good
Good Description and Comments 
Markdown used PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
No Feedback required
Request Description good
